six/modules/c++/cphd/include/cphd/CPHDWriter.h

```diff
@@ -146,75 +314,56 @@ class CPHDWriter
     }
 
 private:
-    void writeMetadata(size_t vbmSize,
-                       size_t cphdSize,
-                       const std::string& classification = "",
-                       const std::string& releaseInfo = "");
+    /*
+     *  Write metadata helper
+     */
+    void writeMetadata(
+        size_t supportSize, // Optional
+        size_t pvpSize,
+        size_t cphdSize);
 
-    void writeVBMData(const sys::ubyte* vbm,
+    /*
+     *  Write pvp helper
+     */
+    void writePVPData(const sys::ubyte* pvpBlock,
                       size_t index);
 
+    /*
+     *  Implementation of write wideband
+     */
     void writeCPHDDataImpl(const sys::ubyte* data,
                            size_t size);
 
-    class DataWriter
-    {
-    public:
-        DataWriter(io::FileOutputStream& stream,
-                   size_t numThreads);
-
-        virtual ~DataWriter();
-
-        virtual void operator()(const sys::ubyte* data,
-                                size_t numElements,
-                                size_t elementSize) = 0;
-
-    protected:
-        io::FileOutputStream& mStream;
-        const size_t mNumThreads;
-    };
-
-    class DataWriterLittleEndian : public DataWriter
-    {
-    public:
-        DataWriterLittleEndian(io::FileOutputStream& stream,
-                               size_t numThreads,
-                               size_t scratchSize);
-
-        virtual void operator()(const sys::ubyte* data,
-                                size_t numElements,
-                                size_t elementSize);
-
-    private:
-        const size_t mScratchSize;
-        const mem::ScopedArray<sys::byte> mScratch;
-    };
-
-    class DataWriterBigEndian : public DataWriter
-    {
-    public:
-        DataWriterBigEndian(io::FileOutputStream& stream,
-                            size_t numThreads);
+    /*
+     *  Implementation of write compressed wideband
+     */
+    void writeCompressedCPHDDataImpl(const sys::ubyte* data,
+                                     size_t channel);
 
-        virtual void operator()(const sys::ubyte* data,
-                                size_t numElements,
-                                size_t elementSize);
-    };
+    /*
+     *  Implementation of write support data
+     */
+    void writeSupportDataImpl(const sys::ubyte* data,
+                              size_t numElements, size_t elementSize);
 
+    //! DataWriter object
     std::auto_ptr<DataWriter> mDataWriter;
 
+    // Book-keeping element
+    //! metadata information
     Metadata mMetadata;
```

Comment:
Can this be const?

Reply:
Well we're updating numPVPBytes in the metadata in CPHDWriter before we write it to the output stream.
I assume the reason for this is if we change/add any pvp parameters we want the metadata to have the correct number of PVP bytes?
I could update the numPVPBytes everytime a pvp is changed/added instead of in CPHDWriter if that makes more sense?

Reply:
Ah, I see. I'm leaning toward saying that if the PVP and metadata are out-of-sync then the caller did something wrong, so go ahead and throw, but also make sure they have a good way to keep them in sync. Like if I'm debugging and the metadata value is silent different than what I passed in, that could be annoying. But whatever you think is best is fine.